hermeto/core/config.py

```diff
@@ -23,40 +23,131 @@
+
+    # This setting exists for legacy use-cases only and must not be relied upon
+    ignore_dependencies_crates: bool = False
+
+
+class YarnSettings(BaseModel):
+    """Settings for Yarn v2+."""
+
+    # This setting exists for legacy use-cases only and must not be relied upon
+    enabled: bool = True
+
+
+class GomodSettings(BaseModel):
+    """Settings for Go modules."""
+
+    proxy_url: str = "https://proxy.golang.org,direct"
+    download_max_tries: int = 5
+    environment_variables: dict[str, str] = {}
+
+
+class HttpSettings(BaseModel):
+    """HTTP-related settings."""
+
+    timeout: int = 300
+
+
+class RuntimeSettings(BaseModel):
+    """General runtime execution settings."""
+
+    subprocess_timeout: int = 3600
+    concurrency_limit: int = 5
+
 
 class Config(BaseSettings):
     """Singleton that provides default configuration for the application process."""
 
     model_config = SettingsConfigDict(
         case_sensitive=False,
+        env_nested_delimiter="__",
         env_prefix=f"{APP_NAME.upper()}_",
         extra="forbid",
     )
 
-    goproxy_url: str = "https://proxy.golang.org,direct"
-    default_environment_variables: dict = {}
-    gomod_download_max_tries: int = 5
-    gomod_strict_vendor: bool = True
-    subprocess_timeout: int = 3600
-
-    # matches aiohttp default timeout:
-    # https://docs.aiohttp.org/en/v3.9.5/client_reference.html#aiohttp.ClientSession
-    requests_timeout: int = 300
-    concurrency_limit: int = 5
-
-    # The flags below are for legacy use-cases compatibility only, must not be
-    # relied upon and will be eventually removed.
-    allow_yarnberry_processing: bool = True
-    ignore_pip_dependencies_crates: bool = False
+    pip: PipSettings = PipSettings()
+    yarn: YarnSettings = YarnSettings()
+    gomod: GomodSettings = GomodSettings()
+    http: HttpSettings = HttpSettings()
+    runtime: RuntimeSettings = RuntimeSettings()
 
     @model_validator(mode="before")
     @classmethod
-    def _print_deprecation_warning(cls, data: Any) -> Any:
-        if "gomod_strict_vendor" in data:
-            log.warning(
-                "The `gomod_strict_vendor` config option is deprecated and will be removed in "
-                f"future versions. Note that it no longer has any effect when set, {APP_NAME} will "
-                "always check the vendored contents and fail if they are not up-to-date."
+    def _normalize_config_structure(cls, data: Any) -> Any:
+        """Normalize config data to the new namespaced structure.
+
+        - Remove deprecated fields with warnings
+        - Migrate legacy flat fields to new namespaced structure
+        """
+        if not isinstance(data, dict):
+            return data
+
+        _remove_gomod_strict_vendor(data)
+
+        for old_key, (namespace, new_key) in _FLAT_FIELD_MIGRATIONS:
+            if old_key in data:
+                _migrate_flat_to_namespace(data, old_key, data.pop(old_key), namespace, new_key)
```

Comment:
Maybe add a `FIXME` commentary with a note to drop  this `flat_to_namespace` logic with Hermeto 1.0